### PR TITLE
ARROW-10206: [C++][Python][FlightRPC] Allow disabling server validation

### PR DESCRIPTION
**Pull request by jduo**
@lidavidm WIP:

- Demonstrate changes for C++ and Java clients.
- Need to determine how to specify gRPC version in C++.
- Need to check over coding style in C++ and Java code.
- Python client changes pending.
- Unit tests pending.
- Does it make sense to deprecate use of SslCredentialsOptions in C++ and use TlsCredentialsOptions for other SSL properties? This is higher in blast radius and uses an experimental API for more, but would be more consistent.

### CODE REVIEW

**Comment by lidavidm** [comment]
Hey James, I'll review this later today hopefully, but just to answer a few questions
- Style in C++ can be done automatically with clang-format, see the [developer guide](https://arrow.apache.org/docs/developers/cpp/development.html). 
- For Python similarly you can use the [flake8 config](https://arrow.apache.org/docs/developers/python.html#coding-style)
- The versions are scattered but as a start see thirdparty.txt: https://github.com/apache/arrow/blob/master/cpp/thirdparty/versions.txt though note you may also have to update CI scripts, Homebrew/Conda packages, etc.
- I'd rather avoid very recent experimental features as that will create a lot of churn for us as gRPC upgrades frequently. It may be worth considering in a few versions if it looks like the API settles down, since I agree it would be nice to consolidate the options + it does bring the ability to do certificate rotation.

**Comment by lidavidm** [comment]
Actually, it seems the relevant changes have been there since early this year. It may be OK to switch fully.

**Review by lidavidm — COMMENTED** [review]
Looks good to me as a start. Can unit tests be added? There are already tests set up to use (self-signed) certs that could serve as a starting point.

**Comment by jduo** [comment]
> Hey James, I'll review this later today hopefully, but just to answer a few questions
> 
> * Style in C++ can be done automatically with clang-format, see the [developer guide](https://arrow.apache.org/docs/developers/cpp/development.html).
> * For Python similarly you can use the [flake8 config](https://arrow.apache.org/docs/developers/python.html#coding-style)
> * The versions are scattered but as a start see thirdparty.txt: https://github.com/apache/arrow/blob/master/cpp/thirdparty/versions.txt though note you may also have to update CI scripts, Homebrew/Conda packages, etc.
> * I'd rather avoid very recent experimental features as that will create a lot of churn for us as gRPC upgrades frequently. It may be worth considering in a few versions if it looks like the API settles down, since I agree it would be nice to consolidate the options + it does bring the ability to do certificate rotation.

Hi @lidavidm .

- I have made the Python edits now and am running them through CI while I get my environment working.

- Unit tests have been added for all three languages.

- I updated basically every file every file I could find that referenced a C++ gRPC version.

**Comment by lidavidm** [comment]
@kszucs - are we ok to bump the gRPC versions like this?

File: cpp/src/arrow/flight/client.h [hunk 1]
```diff
@@ -108,10 +108,14 @@ class ARROW_FLIGHT_EXPORT FlightClientOptions {
...
+
   /// \brief Generic connection options, passed to the underlying
   ///     transport; interpretation is implementation-dependent.
   std::vector<std::pair<std::string, util::variant<int, std::string>>> generic_options;
 
+  /// \brief Use TLS without validating the server certificate. Use with caution.
+  bool disable_server_verification;
+
   /// \brief Get default options.
   static FlightClientOptions Defaults();
```
Comment (lidavidm) [inline cpp/src/arrow/flight/client.h:120]: I believe you're seeing failures since `Defaults` should be updated to set the new flag, else the flag may or may not be initialized.

Comment (jduo) [reply cpp/src/arrow/flight/client.h:122]: I've added an explicitly defined constructor since users might be constructing FlightCredentialsOptions directly. I am surprised that this didn't previously cause problems with the write_size_limit_bytes option though.

Comment (lidavidm) [reply cpp/src/arrow/flight/client.h:122]: The intent was to use Defaults(), at least.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -2463,7 +2463,7 @@ macro(build_grpc)
 endmacro()
 
 if(ARROW_WITH_GRPC)
-  set(ARROW_GRPC_REQUIRED_VERSION "1.17.0")
+  set(ARROW_GRPC_REQUIRED_VERSION "1.32.0")
```
Comment (lidavidm) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:2466]: We should check that Homebrew and Conda, at least, have this packaged.

File: dev/tasks/conda-recipes/.ci_support/linux_aarch64_python3.6.____cpython.yaml [hunk 3]
```diff
@@ -31,7 +31,7 @@ gflags:
 glog:
 - 0.4.0
 grpc_cpp:
-- '1.30'
+- '1.32.0'
```
Comment (lidavidm) [inline dev/tasks/conda-recipes/.ci_support/linux_aarch64_python3.6.____cpython.yaml:34]: Can we avoid pinning a minor version?

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 2 above)
Comment (lidavidm) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:2466]: Both seem fine, fortunately.

**Comment by jduo** [comment]
I've changed this PR to cover just C++ and Python.

**Comment by jduo** [comment]
The Java work has been moved to #8377.

File: cpp/src/arrow/flight/client.cc [hunk 4]
```diff
@@ -34,6 +34,9 @@
 #include <grpc++/grpc++.h>
 #endif
 
+#include "grpc/grpc_security_constants.h"
+#include "grpcpp/security/tls_credentials_options.h"
```
Comment (xhochy) [inline cpp/src/arrow/flight/client.cc:38]: External includes, thus no `"`
```suggestion
#include <grpc/grpc_security_constants.h>
#include <grpcpp/security/tls_credentials_options.h>
```

File: python/manylinux1/scripts/build_re2.sh [hunk 5]
```diff
@@ -31,5 +31,7 @@ make prefix=/usr/local -j${NCORES} install
 
 popd
 
-# Need to remove shared library to make sure the static library is picked up by Arrow
-rm -rf re2-${RE2_VERSION}.tar.gz re2-${RE2_VERSION} /usr/local/lib/libre2.so*
+# Need to remove static library to make sure the shared library is picked up by Arrow
+# The static library fails to link when using gRPC 1.32.
+rm -rf re2-${RE2_VERSION}.tar.gz re2-${RE2_VERSION} /usr/local/lib/libre2.a
```
Comment (xhochy) [inline python/manylinux1/scripts/build_re2.sh:36]: Why was this changed? Manylinux builds should be all static linkage.

Comment (jduo) [reply python/manylinux1/scripts/build_re2.sh:36]: I was seeing linker errors and wasn't aware static linking was a hard requirement for manylinux1.
The linker error was the the string constructor for RE2 was not being defined.

Any thoughts about how to fix this? Flight doesn't use RE2 directly -- gRPC uses RE2. For this type of problem, with GNU Makefiles I'd change the order libraries were linked in to ensure gRPC is fed to the linker before RE2 or add the linker flag to disable stripping of unneeded symbols. I'm not sure about doing this with CMake though.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 2 above)
Comment (nealrichardson) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:2516]: @jduo on the mailing list, you say that it's 1.27 that is required

**Comment by jduo** [comment]
FYI @nealrichardson @pitrou @wesm

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 2 above)
Comment (jduo) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:2516]: Yeah that's correct. The feature is in 1.27, but in a different namespace. When I wrote this I figured for convenience of dev we'd want to use the latest stable release but now see that 1.32 isn't actually available in all our supported environments (MSYS only has 1.29). So instead what I'm planning:
- Define a macro indicating if TlsCredentialsOptions is supported. I'll see if it's possible to auto-detect this.
- Define a macro for the namespace holding TlsCredentialsOptions that is based on version.
- Compile out use of TlsCredentials based on the first macro.
- Revert changes in this PR that update the version the gRPC version 1.32.
- Change the minimum version in the codebase/CI scripts to 1.27 (only if it is already lower).

File: python/manylinux1/scripts/build_grpc.sh [hunk 6]
```diff
@@ -16,8 +16,61 @@
 # specific language governing permissions and limitations
 # under the License.
 
-export GRPC_VERSION="1.29.1"
-export CFLAGS="-fPIC -DGPR_MANYLINUX1=1"
+export GRPC_VERSION="1.32.0"
```
Comment (kszucs) [inline python/manylinux1/scripts/build_grpc.sh:19]: Please update `python/manylinux201x/scripts/build_grpc.sh` as well.

Comment (jduo) [reply python/manylinux1/scripts/build_grpc.sh]: Reverted this actually, since we only need gRPC 1.27 for this functionality to work.

File: python/manylinux1/scripts/build_re2.sh (hunk 5 above)
Comment (jduo) [reply python/manylinux1/scripts/build_re2.sh]: Reverted since we only need gRPC 1.27.

**Comment by jduo** [comment]
I've implemented auto-detection of the namespace of TlsCredentialsOptions now. This should let the code work across multiple gRPC versions. If TlsCredentialsOptions doesn't exist, use of it will get complied out.

Also switched the code to only require 1.27 at the minimum, and reverted changes where gRPC > 1.27
was already being used.

File: cpp/src/arrow/flight/client.cc [hunk 7]
```diff
@@ -845,18 +878,52 @@ class FlightClient::FlightClientImpl {
...
+          struct NoOpTlsAuthorizationCheck
+              : public ge::TlsServerAuthorizationCheckInterface {
+            int Schedule(ge::TlsServerAuthorizationCheckArg* arg) override {
+              arg->set_success(1);
+              arg->set_status(GRPC_STATUS_OK);
+              return 0;
+            }
+          };
+
+          noOpAuthCheck = std::shared_ptr<ge::TlsServerAuthorizationCheckConfig>(
```
Comment (wesm) [inline cpp/src/arrow/flight/client.cc:900]: Use `no_op_auth_check_`

File: cpp/src/arrow/flight/client.cc [hunk 8]
```diff
@@ -845,18 +878,52 @@ class FlightClient::FlightClientImpl {
...
+              new ge::TlsServerAuthorizationCheckConfig(
```
Comment (wesm) [inline cpp/src/arrow/flight/client.cc:901]: Can you use `make_shared` here?

File: cpp/src/arrow/flight/client.cc [hunk 9]
```diff
@@ -845,18 +878,52 @@ class FlightClient::FlightClientImpl {
...
+                  std::shared_ptr<ge::TlsServerAuthorizationCheckInterface>(
+                      new NoOpTlsAuthorizationCheck())));
```
Comment (wesm) [inline cpp/src/arrow/flight/client.cc:903]: possible to use `make_shared`?

File: cpp/src/arrow/flight/client.cc [hunk 10]
```diff
@@ -845,18 +878,52 @@ class FlightClient::FlightClientImpl {
...
+          std::shared_ptr<ge::TlsKeyMaterialsConfig> materials_config(
+              new ge::TlsKeyMaterialsConfig());
```
Comment (wesm) [inline cpp/src/arrow/flight/client.cc:905]: `auto materials_config = std::make_shared<ge::TlsKeyMaterialsConfig>();`?

File: python/pyarrow/_flight.pyx [hunk 11]
```diff
@@ -1106,13 +1114,15 @@ cdef class FlightClient(_Weakrefable):
 
     @classmethod
     def connect(cls, location, tls_root_certs=None, cert_chain=None,
-                private_key=None, override_hostname=None):
+                private_key=None, override_hostname=None,
+                disable_server_verify=None):
```
Comment (wesm) [inline python/pyarrow/_flight.pyx:1118]: Make this consistent? verify -> verification

**Review by wesm — COMMENTED** [review]
This looks okay to me. There are a couple stylistic issues that should be fixed but otherwise IMHO we can merge this

File: python/pyarrow/_flight.pyx (hunk 11 above)
Comment (jduo) [reply python/pyarrow/_flight.pyx:1118]: A few lines lower there's a line which is just disable_server_verification=disable_server_verification which exceeds the line limit. Do you know if there's a way around this?

File: cpp/src/arrow/flight/client.cc (hunk 7 above)
Comment (jduo) [reply cpp/src/arrow/flight/client.cc]: Done.

File: cpp/src/arrow/flight/client.cc (hunk 9 above)
Comment (jduo) [reply cpp/src/arrow/flight/client.cc]: Done

File: cpp/src/arrow/flight/client.cc (hunk 10 above)
Comment (jduo) [reply cpp/src/arrow/flight/client.cc]: Done.

File: python/pyarrow/_flight.pyx (hunk 11 above)
Comment (wesm) [reply python/pyarrow/_flight.pyx:1118]: I'll push a change with alternative formatting that fixes this issue

**Comment by wesm** [comment]
There's some broken stuff for me locally with clang-8

```
In file included from /home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:24:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/grpcpp.h:58:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/server.h:22:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/server_impl.h:37:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_interface.h:31:
/home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/rpc_service_method.h:51:16: error: parameter 'handler_data:' not found in the function declaration [-Werror,-Wdocumentation]
    /// \param handler_data: internal data for the handler.
               ^~~~~~~~~~~~~
/home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/rpc_service_method.h:51:16: note: did you mean 'handler_data'?
    /// \param handler_data: internal data for the handler.
               ^~~~~~~~~~~~~
               handler_data
In file included from /home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:24:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/grpcpp.h:58:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/server.h:22:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/server_impl.h:37:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_interface.h:32:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_context_impl.h:41:
/home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_callback_impl.h:406:18: error: parameter 'ok' not found in the function declaration [-Werror,-Wdocumentation]
  /// \param[in] ok Was it successful? If false, no further write-side operation
                 ^~
/home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_callback_impl.h:412:18: error: parameter 'ok' not found in the function declaration [-Werror,-Wdocumentation]
  /// \param[in] ok Was it successful? If false, no further read-side operation
                 ^~
/home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_callback_impl.h:419:18: error: parameter 'ok' not found in the function declaration [-Werror,-Wdocumentation]
  /// \param[in] ok Was it successful? If false, no further write-side operation
                 ^~
In file included from /home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:24:
In file included from /home/wesm/cpp-toolchain/include/grpcpp/grpcpp.h:59:
/home/wesm/cpp-toolchain/include/grpcpp/server_builder.h:131:14: error: parameter 'selected_port[out]' not found in the function declaration [-Werror,-Wdocumentation]
  /// \param selected_port[out] If not `nullptr`, gets populated with the port
             ^~~~~~~~~~~~~~~~~~
/home/wesm/cpp-toolchain/include/grpcpp/server_builder.h:131:14: note: did you mean 'selected_port'?
  /// \param selected_port[out] If not `nullptr`, gets populated with the port
             ^~~~~~~~~~~~~~~~~~
             selected_port
/home/wesm/cpp-toolchain/include/grpcpp/server_builder.h:207:8: error: declaration is marked with '\deprecated' command but does not have a deprecation attribute [-Werror,-Wdocumentation-deprecated-sync]
  /// \deprecated For backward compatibility.
      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/wesm/cpp-toolchain/include/grpcpp/server_builder.h:208:56: note: add a deprecation attribute to the declaration to silence this warning
  ServerBuilder& SetMaxMessageSize(int max_message_size) {
                                                       ^
                                                         __attribute__((deprecated))
/home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:28:11: error: no member named 'experimental' in namespace 'grpc_impl'; did you mean 'grpc::experimental'?
    const grpc_impl::experimental::TlsCredentialsOptions* options) {
          ^~~~~~~~~~~~~~~~~~~~~~~
          grpc::experimental
/home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_context.h:34:11: note: 'grpc::experimental' declared here
namespace experimental {
          ^
/home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:34:39: error: unused variable 'opt' [-Werror,-Wunused-variable]
  grpc_tls_server_verification_option opt = check(nullptr);
                                      ^
8 errors generated.
ninja: build stopped: subcommand failed.
```

**Comment by jduo** [comment]
> There's some broken stuff for me locally with clang-8, I'm trying to fix
> 
> ```
> In file included from /home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:24:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/grpcpp.h:58:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/server.h:22:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/server_impl.h:37:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_interface.h:31:
> /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/rpc_service_method.h:51:16: error: parameter 'handler_data:' not found in the function declaration [-Werror,-Wdocumentation]
>     /// \param handler_data: internal data for the handler.
>                ^~~~~~~~~~~~~
> /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/rpc_service_method.h:51:16: note: did you mean 'handler_data'?
>     /// \param handler_data: internal data for the handler.
>                ^~~~~~~~~~~~~
>                handler_data
> In file included from /home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:24:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/grpcpp.h:58:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/server.h:22:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/server_impl.h:37:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_interface.h:32:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_context_impl.h:41:
> /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_callback_impl.h:406:18: error: parameter 'ok' not found in the function declaration [-Werror,-Wdocumentation]
>   /// \param[in] ok Was it successful? If false, no further write-side operation
>                  ^~
> /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_callback_impl.h:412:18: error: parameter 'ok' not found in the function declaration [-Werror,-Wdocumentation]
>   /// \param[in] ok Was it successful? If false, no further read-side operation
>                  ^~
> /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_callback_impl.h:419:18: error: parameter 'ok' not found in the function declaration [-Werror,-Wdocumentation]
>   /// \param[in] ok Was it successful? If false, no further write-side operation
>                  ^~
> In file included from /home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:24:
> In file included from /home/wesm/cpp-toolchain/include/grpcpp/grpcpp.h:59:
> /home/wesm/cpp-toolchain/include/grpcpp/server_builder.h:131:14: error: parameter 'selected_port[out]' not found in the function declaration [-Werror,-Wdocumentation]
>   /// \param selected_port[out] If not `nullptr`, gets populated with the port
>              ^~~~~~~~~~~~~~~~~~
> /home/wesm/cpp-toolchain/include/grpcpp/server_builder.h:131:14: note: did you mean 'selected_port'?
>   /// \param selected_port[out] If not `nullptr`, gets populated with the port
>              ^~~~~~~~~~~~~~~~~~
>              selected_port
> /home/wesm/cpp-toolchain/include/grpcpp/server_builder.h:207:8: error: declaration is marked with '\deprecated' command but does not have a deprecation attribute [-Werror,-Wdocumentation-deprecated-sync]
>   /// \deprecated For backward compatibility.
>       ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/wesm/cpp-toolchain/include/grpcpp/server_builder.h:208:56: note: add a deprecation attribute to the declaration to silence this warning
>   ServerBuilder& SetMaxMessageSize(int max_message_size) {
>                                                        ^
>                                                          __attribute__((deprecated))
> /home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:28:11: error: no member named 'experimental' in namespace 'grpc_impl'; did you mean 'grpc::experimental'?
>     const grpc_impl::experimental::TlsCredentialsOptions* options) {
>           ^~~~~~~~~~~~~~~~~~~~~~~
>           grpc::experimental
> /home/wesm/cpp-toolchain/include/grpcpp/impl/codegen/server_context.h:34:11: note: 'grpc::experimental' declared here
> namespace experimental {
>           ^
> /home/wesm/code/arrow/cpp/src/arrow/flight/try_compile/check_tls_opts_127.cc:34:39: error: unused variable 'opt' [-Werror,-Wunused-variable]
>   grpc_tls_server_verification_option opt = check(nullptr);
>                                       ^
> 8 errors generated.
> ninja: build stopped: subcommand failed.
> ```

The files in try_compile shouldn't be built as part of the project -- they should only be invoked by CMake for assessing grpc-cpp version information.

**Comment by wesm** [comment]
> The files in try_compile shouldn't be built as part of the project -- they should only be invoked by CMake for assessing grpc-cpp version information.

Right, but the CMAKE_CXX_FLAGS set at that point are passed along, including `-Werror`. I'm pushing a fix here in a minute

**Comment by jduo** [comment]
> > The files in try_compile shouldn't be built as part of the project -- they should only be invoked by CMake for assessing grpc-cpp version information.
> 
> Right, but the CMAKE_CXX_FLAGS set at that point are passed along, including `-Werror`. I'm pushing a fix here in a minute

Thanks for fixing that Wes.

**Review by lidavidm — APPROVED** [review]
LGTM. I'll follow up on the ML.

File: cpp/src/arrow/flight/client.cc [hunk 12]
```diff
@@ -835,6 +843,31 @@ class GrpcMetadataReader : public FlightMetadataReader {
   std::shared_ptr<std::mutex> read_mutex_;
 };
 
+namespace {
+// Dummy self-signed certificate to be used because TlsCredentials
+// requires root CA certs, even if you are skipping server
+// verification.
+#if defined(GRPC_NAMESPACE_FOR_TLS_CREDENTIALS_OPTIONS)
+const char BLANK_ROOT_PEM[] =
```
Comment (emkornfield) [inline cpp/src/arrow/flight/client.cc:851]: nit: constexpr? or static.

File: cpp/src/arrow/flight/client.h [hunk 13]
```diff
@@ -90,6 +90,8 @@ class ARROW_FLIGHT_EXPORT FlightWriteSizeStatusDetail : public arrow::StatusDeta
 
 class ARROW_FLIGHT_EXPORT FlightClientOptions {
  public:
+  FlightClientOptions();
```
Comment (emkornfield) [inline cpp/src/arrow/flight/client.h:93]: there shouldn't be two ways of getting the default values.  If there needs to be another constructor that takes credentials, then use the static factory pattern.

Comment (jduo) [reply cpp/src/arrow/flight/client.h:93]: The default constructor was already there prior to this patch, it was just being implicitly defined instead of explicitly. I agree in principle that the Defaults() method should be used, however the constructor has already been public and I'm not sure it's worth breaking existing application code.

We're not really consistent about this internally either.
The C++ unit tests make use of both the public constructor and Defaults() method. The Python wrapper uses the public constructor.

File: cpp/src/arrow/flight/client.cc (hunk 12 above)
Comment (jduo) [reply cpp/src/arrow/flight/client.cc]: Made a constexpr

File: cpp/src/arrow/flight/client.cc [hunk 14]
```diff
@@ -845,18 +878,48 @@ class FlightClient::FlightClientImpl {
...
-        if (!options.cert_chain.empty()) {
-          ssl_options.pem_cert_chain = options.cert_chain;
-        }
-        if (!options.private_key.empty()) {
-          ssl_options.pem_private_key = options.private_key;
+      if (scheme == kSchemeGrpcTls) {
+        if (options.disable_server_verification) {
+#if !defined(GRPC_NAMESPACE_FOR_TLS_CREDENTIALS_OPTIONS)
+          return Status::NotImplemented(
+              "Using encryption with server verification is unsupported.");
```
Comment (pitrou) [inline cpp/src/arrow/flight/client.cc:885]: Mention the reason in the error message?

File: cpp/src/arrow/flight/client.cc [hunk 15]
```diff
@@ -845,18 +878,48 @@ class FlightClient::FlightClientImpl {
...
+              arg->set_status(GRPC_STATUS_OK);
+              return 0;
+            }
+          };
+
+          noop_auth_check_ = std::make_shared<ge::TlsServerAuthorizationCheckConfig>(
+              std::make_shared<NoOpTlsAuthorizationCheck>());
+          auto materials_config = std::make_shared<ge::TlsKeyMaterialsConfig>();
+          materials_config->set_pem_root_certs(BLANK_ROOT_PEM);
+          ge::TlsCredentialsOptions tls_options(
```
Comment (pitrou) [inline cpp/src/arrow/flight/client.cc:904]: So gRPC has both `TlsCredentialsOptions` and `SslCredentialOptions`?

Comment (jduo) [reply cpp/src/arrow/flight/client.cc:905]: Yes, but they aren't interchangeable. One works against TlsCredentials and one works against SslCredentials. TlsCredentials is a newer, currently experimental API that can do everything SslCredentials can do and more, such as do certificate reloading and supplying custom callbacks for cert verification.

File: cpp/src/arrow/flight/client.cc (hunk 14 above)
Comment (jduo) [reply cpp/src/arrow/flight/client.cc]: Done.

**Comment by lidavidm** [comment]
Waiting for CI here...

File: cpp/src/arrow/flight/client.h (hunk 13 above)
Comment (lidavidm) [reply cpp/src/arrow/flight/client.h:93]: Given there's already code using the default constructor, I think we can remove it separately: https://issues.apache.org/jira/browse/ARROW-10250

**Review by lidavidm — APPROVED** [review]
Ok, looks like the MacOS build is failing due to an unrelated reason (Cython not being found) and other PRs (and master) are failing in the same way. (Travis still pending...)

**Comment by wesm** [comment]
thanks everyone
